### PR TITLE
DLCQuest: progression coin bundle update

### PR DESCRIPTION
**Pull request by axe-y**
## What is this fixing or adding?
small refactoring, changed some item into progressive weapon, speed up generation by changing most coin bundle into useful and not progression, and a small Aprils fool update

## How was this tested?
generation and webhost to verifies it run as expected

### CODE REVIEW

**Comment by NewSoupVi** [comment]
Just for the record, there is a distinction between progress**ion** and progress**ive**, you should try to use the proper word for each case because otherwise it can be very confusing :)

(Progression: This item is logical
Progressive: There are multiple "stages"/"upgrades" of the same item)

**Comment by agilbert1412** [comment]
> Just for the record, there is a distinction between progress**ion** and progress**ive**, you should try to use the proper word for each case because otherwise it can be very confusing :)
> 
> (Progression: This item is logically relevant Progressive: There are multiple "stages"/"upgrades" of the same item)
> 
> I believe in this case you mean that

In this case though, I believe this update introduces changes to both

File: worlds/dlcquest/data/items.csv [hunk 1]
```diff
@@ -28,21 +28,22 @@ id,name,classification,groups
...
 31,High Definition Next Gen Pack,filler,"DLC,Freemium"
 32,Increased HP Pack,useful,"DLC,Freemium"
 33,Removed Ads Pack,filler,"DLC,Freemium"
 34,Big Sword Pack,progression,"DLC,Freemium"
 35,Really Big Sword Pack,progression,"DLC,Freemium"
 36,Unfathomable Sword Pack,progression,"DLC,Freemium"
-37,Gun,progression,"Item,DLCQuest"
-38,Sword,progression,"Item,DLCQuest"
-39,Wooden Sword,progression,"Item,Freemium"
+37,Progressive Weapon Basic,progression,"Item,Twice,DLCQuest"
```
Comment (agilbert1412) [inline worlds/dlcquest/data/items.csv:38]: I would recommend naming these using the same pattern as the coin bundles
DLC Quest: Progressive Weapon
Live Freemium or Die: Progressive Weapon

File: worlds/dlcquest/Items.py [hunk 2]
```diff
@@ -94,40 +97,65 @@ def create_trap_items(world, World_Options: Options.DLCQuestOptions, trap_needed
     return traps
 
 
-def create_items(world, World_Options: Options.DLCQuestOptions, locations_count: int, random: Random):
+def create_items(world, world_options: Options.DLCQuestOptions, locations_count: int, random: Random):
     created_items = []
-    if World_Options.campaign == Options.Campaign.option_basic or World_Options.campaign == Options.Campaign.option_both:
```
Comment (agilbert1412) [inline worlds/dlcquest/Items.py:99]: You can remove the "Options." prefix using the new system and slightly different imports. It'll considerably reduce bloat in this whole file (and others, maybe?)

File: worlds/dlcquest/Items.py [hunk 3]
```diff
@@ -94,40 +97,65 @@ def create_trap_items(world, World_Options: Options.DLCQuestOptions, trap_needed
...
+    created_items += trap_items
+
+    return created_items
+
+
+def create_items_lfod(world_options, created_items, world):
+    for item in items_by_group[Group.Freemium]:
+        if item.has_any_group(Group.DLC):
+            created_items.append(world.create_item(item))
+        if item.has_any_group(Group.Item) and world_options.item_shuffle == Options.ItemShuffle.option_shuffled:
```
Comment (agilbert1412) [inline worlds/dlcquest/Items.py:119]: (Can't believe I'm saying this, feel free to ignore it, it's painful to even think)
You can improve performance by doing the check for item shuffle outside of the loop
ughh I feel dirty now.

File: worlds/dlcquest/Items.py [hunk 4]
```diff
@@ -94,40 +97,65 @@ def create_trap_items(world, World_Options: Options.DLCQuestOptions, trap_needed
...
+    if world_options.coinsanity == Options.CoinSanity.option_coin:
+        if world_options.coinbundlequantity == -1:
+            create_coin_piece(created_items, world, 825, 250, Group.DLCQuest)
+            return
+        create_coin(world_options, created_items, world, 825, 250, Group.DLCQuest)
+
+
+def create_coin(world_options, created_items, world, total_coin, required_coin, group):
+    coin_bundle_required = math.ceil(required_coin / world_options.coinbundlequantity)
+    coin_bundle_useful = math.ceil((total_coin - coin_bundle_required * world_options.coinbundlequantity) / world_options.coinbundlequantity)
```
Comment (agilbert1412) [inline worlds/dlcquest/Items.py:147]:
```py
total_coin_bundles = math.ceil(total_coin / world_options.coinbundlequantity)
coin_bundle_required = math.ceil(required_coin / world_options.coinbundlequantity)
coin_bundle_useful = total_coin_bundles - coin_bundle_required
```

File: worlds/dlcquest/Items.py [hunk 5]
```diff
@@ -94,40 +97,65 @@ def create_trap_items(world, World_Options: Options.DLCQuestOptions, trap_needed
...
+    for item in items_by_group[group]:
```
Comment (agilbert1412) [inline worlds/dlcquest/Items.py:148]: looping over all items to find the one coin?

File: worlds/dlcquest/Items.py [hunk 6]
```diff
@@ -94,40 +97,65 @@ def create_trap_items(world, World_Options: Options.DLCQuestOptions, trap_needed
...
-                        created_items.append(world.create_item(item))
-
-    trap_items = create_trap_items(world, World_Options, locations_count - len(created_items), random)
-    created_items += trap_items
+            for i in range(coin_bundle_useful):
+                created_items.append(DLCQuestItem(item.name, ItemClassification.useful, item.code, world.player))
 
-    return created_items
+def create_coin_piece(created_items, world, total_coin, require_coin, group):
+    for item in items_by_group[group]:
```
Comment (agilbert1412) [inline worlds/dlcquest/Items.py:156]: looping over all items to find the one coin piece?

**Review by agilbert1412 — APPROVED** [review]

File: worlds/dlcquest/Options.py [hunk 7]
```diff
@@ -48,6 +49,14 @@ class CoinSanityRange(NamedRange):
         "normal": 20,
         "high": 50,
     }
+    if datetime.datetime.today().month == 4:
```
[the remark this answers is not in the document]
Comment (axe-y) [reply worlds/dlcquest/Options.py:52]: will test this tonight

File: worlds/dlcquest/data/items.csv (hunk 1 above)
Comment (axe-y) [reply worlds/dlcquest/data/items.csv]: done

File: worlds/dlcquest/Options.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (axe-y) [reply worlds/dlcquest/Options.py:52]: I did test I don't think it is but I am not sure if I can do anything

[the remark this answers is not in the document]
Comment (axe-y) [reply worlds/dlcquest/Options.py:52]: So more test and the one for factorio doesn't change so I don't know what happen but maybe berzerker reset the server for Aprils

File: worlds/dlcquest/data/items.csv [hunk 8]
```diff
@@ -28,21 +28,22 @@ id,name,classification,groups
...
 30,Season Pass,progression,"DLC,Freemium"
 31,High Definition Next Gen Pack,filler,"DLC,Freemium"
 32,Increased HP Pack,useful,"DLC,Freemium"
 33,Removed Ads Pack,filler,"DLC,Freemium"
 34,Big Sword Pack,progression,"DLC,Freemium"
 35,Really Big Sword Pack,progression,"DLC,Freemium"
 36,Unfathomable Sword Pack,progression,"DLC,Freemium"
-37,Gun,progression,"Item,DLCQuest"
-38,Sword,progression,"Item,DLCQuest"
-39,Wooden Sword,progression,"Item,Freemium"
```
Comment (agilbert1412) [inline worlds/dlcquest/data/items.csv:41]: I would personally keep the old items, and mark them as deprecated so they don't get generated, for the sake of backward compatibility.

But even if you do delete them, please don't re-use the same IDs for the new items, as that can cause issues on trackers, text clients, hints, etc.

**Review by agilbert1412 — CHANGES_REQUESTED** [review]

File: worlds/dlcquest/Locations.py [hunk 9]
```diff
@@ -76,3 +76,14 @@ class DLCQuestLocation(Location):
 for i in range(1, 890):
     item_coin_freemium = f"Live Freemium or Die: {i} Coin"
     location_table[item_coin_freemium] = offset + 825 + 58 + i
+
+
+offset_special = 3829200000
+
+for i in range(1, 8251):
+    item_coin_piece = f"DLC Quest: {i / 10} Coin Piece"
```
Comment (agilbert1412) [inline worlds/dlcquest/Locations.py:84]: If they are called "X Coin Piece", X should be a whole number from 1 to N * 10, not a decimal number from 0.1 to N

**Review by agilbert1412 — CHANGES_REQUESTED** [review]

File: worlds/dlcquest/Items.py (hunk 2 above)
Comment (axe-y) [reply worlds/dlcquest/Items.py:99]: I don't want to switch them all (maybe another time) and I don't want both system at the same time

File: worlds/dlcquest/Items.py [hunk 10]
```diff
@@ -94,40 +97,65 @@ def create_trap_items(world, World_Options: Options.DLCQuestOptions, trap_needed
...
+        if item.has_any_group(Group.Piece):
+            for i in range(require_coin*10):
+                created_items.append(world.create_item(item))
+            for i in range(total_coin - require_coin):
```
[the remark this answers is not in the document]
Comment (axe-y) [reply worlds/dlcquest/Items.py]: fix

**Review by agilbert1412 — APPROVED** [review]

**Review by agilbert1412 — APPROVED** [review]

**Review by Berserker66 — APPROVED** [review]

**Pull request merged**